oximeter/db/src/native/mod.rs

```diff
@@ -0,0 +1,197 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+//
+// Copyright 2024 Oxide Computer Company
+
+//! Low-level native client to the ClickHouse database.
+//!
+//! The ClickHouse server offers a number of client interfaces. We continue to
+//! use JSON-over-HTTP today, because it's so easy. The server communicates with
+//! the official client using what they call the "Native" protocol over TCP
+//! sockets. This module implements a prototype client able to speak this
+//! protocol directly.
+//!
+//! # Rationale
+//!
+//! For a long time, we've made do by talking to ClickHouse over HTTP, and
+//! transferring data as JSON rows. This works, but has a number of drawbacks.
+//! It incurs a lot of serialization overhead, both for the client and the
+//! server. JSON is pretty verbose, and the server also needs to do a lot of
+//! work to convert the native columnar formats into row-based outputs. All of
+//! that goes away with the native protocol, where both peers can do little more
+//! than memcpy the columns over the network.
+//!
+//! The HTTP interface also incurs a ot of additional _latency_. We ask the
```

Comment:
It will be so great to get rid of this behavior. When I read the original http code I got scared ;)